FirebaseRemoteConfig/SwiftNew/ConfigDBManager.swift

```diff
@@ -0,0 +1,374 @@
+import Foundation
+import SQLite3
+
+@objc public enum UpdateOption: Int {
+  case applyTime
+  case defaultTime
+  case fetchStatus
+}
+
+/// Column names in metadata table
+let RCNKeyBundleIdentifier = "bundle_identifier"
+let RCNKeyNamespace = "namespace"
+let RCNKeyFetchTime = "fetch_time"
+let RCNKeyDigestPerNamespace = "digest_per_ns"
+let RCNKeyDeviceContext = "device_context"
+let RCNKeyAppContext = "app_context"
+let RCNKeySuccessFetchTime = "success_fetch_time"
+let RCNKeyFailureFetchTime = "failure_fetch_time"
+let RCNKeyLastFetchStatus = "last_fetch_status"
+let RCNKeyLastFetchError = "last_fetch_error"
+let RCNKeyLastApplyTime = "last_apply_time"
+let RCNKeyLastSetDefaultsTime = "last_set_defaults_time"
+
+/// SQLite file name in versions 0, 1 and 2.
+private let RCNDatabaseName = "RemoteConfig.sqlite3"
+/// The storage sub-directory that the Remote Config database resides in.
+private let RCNRemoteConfigStorageSubDirectory = "Google/RemoteConfig"
+
+// TODO: Delete all publics and objc's
+
+/// Persist config data in sqlite database on device. Managing data read/write from/to database.
+@objc(RCNConfigDBManager)
+open class ConfigDBManager: NSObject {
+  /// Shared Singleton Instance
+  @objc public static let sharedInstance = ConfigDBManager()
+
+  private let databaseActor: DatabaseActor
+
+  @objc public var isNewDatabase: Bool = false
+
+  @objc public init(dbPath: String = remoteConfigPathForDatabase()) {
+    databaseActor = DatabaseActor(dbPath: dbPath)
+    super.init()
+  }
+
+  /// Returns the current version of the Remote Config database.
+  public static func remoteConfigPathForDatabase() -> String {
+    #if os(tvOS)
+      let dirPaths = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)
+    #else
+      let dirPaths = NSSearchPathForDirectoriesInDomains(.applicationSupportDirectory,
+                                                         .userDomainMask, true)
+    #endif
+    let storageDirPath = dirPaths[0]
+    let dbPath = URL(fileURLWithPath: storageDirPath)
+      .appendingPathComponent(RCNRemoteConfigStorageSubDirectory)
+      .appendingPathComponent(RCNDatabaseName).path
+    return dbPath
+  }
+
+  private func RemoteConfigPathForOldDatabaseV0() -> String {
+    let dirPaths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
+    let docPath = dirPaths[0]
+    return URL(fileURLWithPath: docPath).appendingPathComponent(RCNDatabaseName).path
+  }
+
+  // MARK: - Insert
+
+  @objc public
+  func insertMetadataTable(withValues columnNameToValue: [String: Any],
+                           completionHandler handler: ((Bool, [String: AnyHashable]?) -> Void)? =
+                             nil) {
+    Task { // Use Task to call the actor method asynchronously
+      let success = await self.databaseActor.insertMetadataTable(withValues: columnNameToValue)
+      if let handler {
+        handler(success, nil) // Call the completion handler
```

Comment:
The DispatchQueue.main can cause a deadlock. This can be converted to an `async` function when we move all the clients to Swift. Note: nothing in this file is public.